mesonbuild/scripts/run_tool.py

```diff
@@ -3,17 +3,86 @@
 
 from __future__ import annotations
 
-import itertools
+import asyncio
+import asyncio.subprocess
 import fnmatch
-import concurrent.futures
+import itertools
+import signal
+import sys
 from pathlib import Path
 
+from .. import mlog
 from ..compilers import lang_suffixes
 from ..mesonlib import quiet_git
+from ..mtest import sh_quote, determine_worker_count, complete_all
 import typing as T
 
-if T.TYPE_CHECKING:
-    import subprocess
+Info = T.TypeVar("Info")
+
+async def run_with_buffered_output(cmdlist: T.List[str]) -> int:
+    """Run the command in cmdlist, buffering the output so that it is
+       not mixed for multiple child processes.  Kill the child on
+       cancellation."""
+    quoted_cmdline = ' '.join(sh_quote(x) for x in cmdlist)
+    p: T.Optional[asyncio.subprocess.Process] = None
+    try:
+        p = await asyncio.create_subprocess_exec(*cmdlist,
+                                                 stdin=asyncio.subprocess.DEVNULL,
+                                                 stdout=asyncio.subprocess.PIPE,
+                                                 stderr=asyncio.subprocess.STDOUT)
+        stdo, _ = await p.communicate()
+    except FileNotFoundError as e:
+        print(mlog.blue('>>>'), quoted_cmdline, file=sys.stderr)
+        print(mlog.red('not found:'), e.filename, file=sys.stderr)
+        return 1
+    except asyncio.CancelledError:
+        if p:
+            p.kill()
+            await p.wait()
+            return p.returncode or 1
+        else:
+            return 0
+
+    if stdo:
+        print(mlog.blue('>>>'), quoted_cmdline, flush=True)
+        sys.stdout.buffer.write(stdo)
+    return p.returncode
+
+async def _run_workers(infos: T.Iterable[Info],
+                       fn: T.Callable[[Info], T.Iterable[T.Coroutine[None, None, int]]]) -> int:
+    futures: T.List[asyncio.Future[int]] = []
+    semaphore = asyncio.Semaphore(determine_worker_count())
```

Comment:
determine_worker_count() is highly specific to `meson test`. It is in fact identical to multiprocessing.cpu_count() except when the environment variable `$MESON_TESTTHREADS` is exported, and it is obviously incorrect behavior for `ninja clang-format` to be controlled by a "test threads" setting.

Reply:
It is clearly useful for more than `meson test` though. I will add support for MESON_THREADS, with extra fallback to MESON_TESTTHREADS for `meson test` only.